File: AdaptiveTitForTat.java
class AdaptiveTitForTat extends Prisoner {
    /*Works as follows
    An adaption rate r is used to compute a continuous variable ‘world’ according to the history moves of the opponent. The logic of the strategy is shown as below.

    If (opponent played C in the last cycle) then
    world = world + r*(1-world), r is the adaptation rate
    else
    world = world + r*(0-world)
    If (world >= 0.5) play C, else play D
    */

    private double r;
    private double world;

    public AdaptiveTitForTat(double r) {
        //Constructor where 'r' is the adaptation rate 0 < r <= 1
        super("Adaptive (r=" + r + ")");
        this.r = r;
        reset();
    }

    public int play() {
        //Make and return a move
        if (lastMove != -1) {
            //Not the first round of the game
            if (opponent.getLastMove() == 1) {
                world += r*(1-world);
            }
            else {
                world += r * (0-world);
            }
        }
        return (world >= 0.5) ? 1 : 0;
    }

    public void reset() {
        world = 1.0;
        super.reset();
    }

    public Prisoner clone() {
        //Return an offspring (for evolution tournament)
        return new AdaptiveTitForTat(r);
    }
}

File: prisoner/AdaptiveTitForTat.kt
package de.freyama.prisoners_dilemma.prisoner

/**
 * An AdaptiveTitForTat Prisoner works by adapting to the world around them.
 * The strategy works as follows;
 *
 * An adaption rate `r` is provided at initialisation.
 * This rate is used to compute a continuous value `world` according to the moves the opponent makes.
 * The logic of this calculation is as follows;
 *
 * if opponent cooperated in the last bout
 * then
 *   world += r * (1 - world)
 * else
 *   world += r * (0 - world)
 * end
 *
 * Then, using the value of `world`, the Prisoner determines what move to make.
 * They will cooperate if `world` >= 0.5, otherwise they will defect
 */
class AdaptiveTitForTat(private val adaption: Double) : Prisoner("TitForTat (ADAPTIVE) (r=$adaption)") {
    private var world: Double = 1.0

    /**
     * Make a move be adapting to the world after the previous move and using the new state to pick a move.
     */
    override fun play(): Boolean {
        adapt()
        val move = this.world >= 0.5
        this.lastMove = move
        return move
    }

    /**
     * Adapt to the world using the opponent's last move
     */
    private fun adapt() {
        val lastMove = this.opponent?.lastMove
        if (lastMove == null) {
            return
        }

        if (lastMove) {
            // Cooperated, increase world value
            this.world += this.adaption * (1 - world)
        }
        else {
            // Defected, decrease world value
            this.world += this.adaption * (0 - world)
        }
    }

    /**
     * Override reset to set the world value back to 1.0
     */
    override fun reset() {
        this.world = 1.0
        super.reset()
    }
}
